File: sudoku.ml
open Base
open Stdio

let sudoku_size = 9

let fill_line puzzle line_count line =
  for i = 0 to (sudoku_size - 1) do
    puzzle.(line_count).(i) <- line.(i)
  done

let rec read_puzzle puzzle line_count = 
  if line_count = sudoku_size then Some puzzle else
    let line = In_channel.input_line In_channel.stdin in
    match line with
    | None -> None
    | Some str -> let line_arr = (Array.of_list (List.map ~f:Int.of_string (String.split ~on:' ' str))) in
                    if Array.length line_arr <> sudoku_size then None
                    else (fill_line puzzle line_count line_arr; read_puzzle puzzle (line_count + 1))

let print_puzzle puzzle =
  for i = 0 to (sudoku_size - 1) do
    for j = 0 to (sudoku_size - 1) do
      printf "%d " puzzle.(i).(j)
    done;
    printf "\n"
  done

let rec row_conflict puzzle row i nums =
  if i = sudoku_size then false
  else if puzzle.(row).(i) <> 0 && Set.mem nums puzzle.(row).(i) then true
  else row_conflict puzzle row (i + 1) (Set.add nums puzzle.(row).(i))

let rec col_conflict puzzle col i nums =
  if i = sudoku_size then false
  else if puzzle.(i).(col) <> 0 && Set.mem nums puzzle.(i).(col) then true
  else col_conflict puzzle col (i + 1) (Set.add nums puzzle.(i).(col))

let rec box_conflict puzzle box i nums =
  if i = sudoku_size then false
  else
    let row = (i / 3) + (3 * (box / 3)) in
    let col = (i % 3) + (3 * (box % 3)) in
    if puzzle.(row).(col) <> 0 && Set.mem nums puzzle.(row).(col) then true
    else box_conflict puzzle box (i + 1) (Set.add nums puzzle.(row).(col))

let has_conflict puzzle row col =
  row_conflict puzzle row 0 (Set.empty (module Int)) ||
  col_conflict puzzle col 0 (Set.empty (module Int)) ||
  box_conflict puzzle ((3 * (row / 3)) + (col / 3)) 0 (Set.empty (module Int))

let rec solve puzzle i =
  if i = (sudoku_size * sudoku_size) then Some puzzle
  else
    let row = i / sudoku_size in
    let col = i % sudoku_size in
    if puzzle.(row).(col) <> 0 then
      solve puzzle (i + 1)
    else
      let solution = ref None in
      let num = ref 1 in
      while (match !solution with None -> true | Some _ -> false) && (!num <= 9) do
        puzzle.(row).(col) <- !num;
        (if not (has_conflict puzzle row col) then
          solution := solve puzzle (i + 1));
        num := !num + 1
      done;
      (if (match !solution with None -> true | Some _ -> false) then
        puzzle.(row).(col) <- 0);
      !solution

let () =
  let puzzle_opt = read_puzzle (Array.make_matrix ~dimx:sudoku_size ~dimy:sudoku_size 0) 0 in
  match puzzle_opt with
    | None -> printf "Error\n"
    | Some puzzle ->
      printf "\n";
      match solve puzzle 0 with
      | None -> printf "No Solution\n"
      | Some solution -> print_puzzle solution
  

File: cpp_version/sudoku.cpp
#include <iostream>
#include <vector>
#include <set>

using std::set;
using std::vector;
using std::cin;
using std::cout;
using std::endl;

constexpr int SUDOKU_SIZE = 9;

bool row_conflict(vector<vector<int>>& puzzle, int row) {
    set<int> nums;
    for (int i = 0; i < SUDOKU_SIZE; ++i) {
        if (nums.count(puzzle[row][i]) > 0) {
            return true;
        }
        if (puzzle[row][i] != 0) {
            nums.insert(puzzle[row][i]);
        }
    }
    return false;
}

bool col_conflict(vector<vector<int>>& puzzle, int col) {
    set<int> nums;
    for (int i = 0; i < SUDOKU_SIZE; ++i) {
        if (nums.count(puzzle[i][col]) > 0) {
            return true;
        }
        if (puzzle[i][col] != 0) {
            nums.insert(puzzle[i][col]);
        }
    }
    return false;
}

bool box_conflict(vector<vector<int>>& puzzle, int row, int col) {
    set<int> nums;
    constexpr int BOX_SIZE = SUDOKU_SIZE / 3;
    int box_row = row - (row % BOX_SIZE);
    int box_col = col - (col % BOX_SIZE);
    for (int i = box_row; i < box_row + BOX_SIZE; ++i) {
        for (int j = box_col; j < box_col + BOX_SIZE; ++j) {
            if (nums.count(puzzle[i][j]) > 0) {
                return true;
            }
            if (puzzle[i][j] != 0) {
                nums.insert(puzzle[i][j]);
            }
        }
    }
    return false;
}

bool has_conflict(vector<vector<int>>& puzzle, int row, int col) {
    return row_conflict(puzzle, row) || col_conflict(puzzle, col) || box_conflict(puzzle, row, col);
}

void next(int& row, int& col) {
    int temp = (row * SUDOKU_SIZE) + col + 1;
    row = temp / SUDOKU_SIZE;
    col = temp % SUDOKU_SIZE;
}

void prev(int& row, int& col) {
    int temp = (row * SUDOKU_SIZE) + col - 1;
    row = temp / SUDOKU_SIZE;
    col = temp % SUDOKU_SIZE;
}

// returns a bool indicating if the puzzle has a solution
bool solve(vector<vector<int>>& puzzle, int row, int col) {
    if (row == SUDOKU_SIZE) {
        return true;
    }
    if (puzzle[row][col] != 0) {
        next(row, col);
        bool ans = solve(puzzle, row, col);
        prev(row, col);
        return ans;
    }
    for (int i = 1; i <= SUDOKU_SIZE; ++i) {
        puzzle[row][col] = i;
        if (!has_conflict(puzzle, row, col)) {
            next(row, col);
            bool solvable = solve(puzzle, row, col);
            prev(row, col);
            if (solvable) {
                return solvable;
            }
        }
    }
    puzzle[row][col] = 0;
    return false;
}

int main() {
    // read in puzzle
    vector<vector<int>> puzzle = vector<vector<int>>(SUDOKU_SIZE, vector<int>(SUDOKU_SIZE, 0));
    for (int i = 0; i < SUDOKU_SIZE; ++i) {
        for (int j = 0; j < SUDOKU_SIZE; ++j) {
            cin >> puzzle[i][j];
        }
    }
    cout << endl;

    // solve puzzle
    bool solvable = solve(puzzle, 0, 0);
    if (!solvable) {
        cout << "No Solution" << endl;
        return 0;
    }

    // output puzzle
    for (vector<int>& row : puzzle) {
        for (int num : row) {
            cout << num << " ";
        }
        cout << endl;
    }
}